# Project 4

## Part 1: Getting the Data

In [ ]:
!pip install folium

In [2]:
import folium
import requests
import pandas

arrest_table = pandas.read_csv("https://cmsc320.github.io/files/BPD_Arrests.csv")

arrest_table = arrest_table[pandas.notnull(arrest_table["Location 1"])]

arrest_table["lat"], arrest_table["long"] = arrest_table["Location 1"].str.split(",").str
arrest_table["lat"] = arrest_table["lat"].str.replace("(", "").astype(float)
arrest_table["long"] = arrest_table["long"].str.replace(")", "").astype(float)

<ipython-input-2-036a81c3980c>:9: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  arrest_table["lat"], arrest_table["long"] = arrest_table["Location 1"].str.split(",").str


In [3]:
arrest_table

arrest  age race sex  arrestDate arrestTime      arrestLocation  \
1       11127013.0   37    B   M  01/01/2011   00:01:00    2000 Wilkens Ave   
2       11126887.0   46    B   M  01/01/2011   00:01:00   2800 Mayfield Ave   
3       11126873.0   50    B   M  01/01/2011   00:04:00   2100 Ashburton St   
4       11126968.0   33    B   M  01/01/2011   00:05:00     4000 Wilsby Ave   
5       11127041.0   41    B   M  01/01/2011   00:05:00    2900 Spellman Rd   
...            ...  ...  ...  ..         ...        ...                 ...   
104522  13610388.0   32    B   M  12/31/2012   23:00:00     3300 Pulaski St   
104523  13610389.0   27    B   M  12/31/2012   23:10:00  400 W Baltimore St   
104524  13610377.0   34    B   M  12/31/2012   23:30:00      3800 Belair Rd   
104525  13610383.0   38    B   M  12/31/2012   23:40:00     2000 Mckean Ave   
104527  13610460.0   43    B   F  12/31/2012   23:55:00    4200 Towanda Ave   

              incidentOffense          incidentLocation  charge  \
1                    79-Other  Wilkens Av & S Payson St  1 1425   
2             Unknown Offense                       NaN     NaN   
3                    79-Other         2100 Ashburton St  1 1106   
4             Unknown Offense         1700 Aliceanna St     NaN   
5       81-Recovered Property           2900 Spelman Rd  1 1425   
...                       ...                       ...     ...   
104522               79-Other           3300 Pulaski Hw  1 5212   
104523        Unknown Offense                       NaN  2 0055   
104524      4E-Common Assault            3600 Belair Rd  1 1415   
104525           87-Narcotics            1700 Mckean Av  2A0696   
104527   4C-Agg. Asslt.- Oth.           4200 Towanda Av  1 1415   

                                  chargeDescription      district   post  \
1       Reckless Endangerment || Hand Gun Violation      SOUTHERN  934.0   
2                                    Unknown Charge  NORTHEASTERN  415.0   
3             Reg Firearm:Illegal Possession || Hgv       WESTERN  735.0   
4                                    Unknown Charge      NORTHERN  525.0   
5        Reckless Endangerment || Handgun Violation      SOUTHERN  924.0   
...                                             ...           ...    ...   
104522       Handgun On Person || Handgun Violation  SOUTHEASTERN  223.0   
104523     Fail Obey Renble/Lawfl || Disorderly/Fto       CENTRAL  113.0   
104524                  Asslt-Sec Degree || Assault  NORTHEASTERN  415.0   
104525  Att-Cds Manuf/Dist-Narc || Distribution Cds       WESTERN  733.0   
104527       Asslt-Sec Degree || 1St Degree Assault  NORTHWESTERN  613.0   

                  neighborhood                       Location 1        lat  \
1             Carrollton Ridge  (39.2814026274, -76.6483635135)  39.281403   
2                Belair-Edison  (39.3227699160, -76.5735750473)  39.322770   
3       Panway/Braddish Avenue  (39.3117196723, -76.6623546313)  39.311720   
4                     Pen Lucy  (39.3382885254, -76.6045667070)  39.338289   
5                  Cherry Hill  (39.2449886230, -76.6273582432)  39.244989   
...                        ...                              ...        ...   
104522   Ellwood Park/Monument  (39.2958396988, -76.5712467336)  39.295840   
104523                Downtown  (39.2893323126, -76.6210021717)  39.289332   
104524           Belair-Edison  (39.3252613570, -76.5689030474)  39.325261   
104525               Mondawmin  (39.3116837460, -76.6475011849)  39.311684   
104527        Towanda-Grantley  (39.3361054128, -76.6671512244)  39.336105   

             long  
1      -76.648364  
2      -76.573575  
3      -76.662355  
4      -76.604567  
5      -76.627358  
...           ...  
104522 -76.571247  
104523 -76.621002  
104524 -76.568903  
104525 -76.647501  
104527 -76.667151  

[63892 rows x 17 columns]

## Part 2: Making a Map

In [4]:
map_osm = folium.Map(location=[39.29, -76.61], zoom_start=11)
map_osm

## Part 3: Combining Parts 1 and 2

First, there are too many arrest to plot all of them. In order to reduce the number, we could look only at a portion of them -- say, the most recent two weeks of arrests. I chose to do this because ideally, the most recent data will be most reflective of the current data. (I fine-tuned this value until the visualization stopped making my computer lag.)

In [4]:
import datetime

arrest_table["arrestDateObj"] = arrest_table["arrestDate"].apply(lambda row: datetime.datetime.strptime(row,"%m/%d/%Y"))
#arrest_table

In [5]:
limitDate = datetime.date(2012, 12, 17)
arrest_table_limited = arrest_table[arrest_table['arrestDateObj'].dt.date > limitDate]
arrest_table_limited

arrest  age race sex  arrestDate arrestTime  \
103194  12604875.0   22    B   M  12/18/2012   00:14:00   
103195  12604908.0   31    B   M  12/18/2012   01:30:00   
103196  12604903.0   30    W   F  12/18/2012   02:02:00   
103197         NaN   32    W   F  12/18/2012   02:10:00   
103198  12604895.0   22    W   M  12/18/2012   02:20:00   
...            ...  ...  ...  ..         ...        ...   
104522  13610388.0   32    B   M  12/31/2012   23:00:00   
104523  13610389.0   27    B   M  12/31/2012   23:10:00   
104524  13610377.0   34    B   M  12/31/2012   23:30:00   
104525  13610383.0   38    B   M  12/31/2012   23:40:00   
104527  13610460.0   43    B   F  12/31/2012   23:55:00   

                  arrestLocation       incidentOffense incidentLocation  \
103194          800 N Calhoun St       Unknown Offense              NaN   
103195          4900 Clifton Ave          87-Narcotics  4900 Clifton Av   
103196          3300 Pulaski Hwy          87-Narcotics  3300 Pulaski Hw   
103197  1400 W Old Coldspring Ln       Unknown Offense              NaN   
103198          3300 Pulaski Hwy          87-Narcotics  3300 Pulaski Hw   
...                          ...                   ...              ...   
104522           3300 Pulaski St              79-Other  3300 Pulaski Hw   
104523        400 W Baltimore St       Unknown Offense              NaN   
104524            3800 Belair Rd     4E-Common Assault   3600 Belair Rd   
104525           2000 Mckean Ave          87-Narcotics   1700 Mckean Av   
104527          4200 Towanda Ave  4C-Agg. Asslt.- Oth.  4200 Towanda Av   

        charge                                  chargeDescription  \
103194     NaN                                     Unknown Charge   
103195  1 0573        Cds: Possession-Marihuana || Poss Marijuana   
103196  4 3550           Cds:Possess-Not Marihuana || Poss Heroin   
103197     NaN                                     Unknown Charge   
103198  4 3550  Cds:Possess-Not Marihuana || Distribution Cocaine   
...        ...                                                ...   
104522  1 5212             Handgun On Person || Handgun Violation   
104523  2 0055           Fail Obey Renble/Lawfl || Disorderly/Fto   
104524  1 1415                        Asslt-Sec Degree || Assault   
104525  2A0696        Att-Cds Manuf/Dist-Narc || Distribution Cds   
104527  1 1415             Asslt-Sec Degree || 1St Degree Assault   

            district   post           neighborhood  \
103194       WESTERN  713.0            Harlem Park   
103195  SOUTHWESTERN  821.0           Franklintown   
103196  SOUTHEASTERN  223.0  Ellwood Park/Monument   
103197      NORTHERN  536.0               Medfield   
103198  SOUTHEASTERN  223.0  Ellwood Park/Monument   
...              ...    ...                    ...   
104522  SOUTHEASTERN  223.0  Ellwood Park/Monument   
104523       CENTRAL  113.0               Downtown   
104524  NORTHEASTERN  415.0          Belair-Edison   
104525       WESTERN  733.0              Mondawmin   
104527  NORTHWESTERN  613.0       Towanda-Grantley   

                             Location 1        lat       long arrestDateObj  
103194  (39.2978595101, -76.6401023080)  39.297860 -76.640102    2012-12-18  
103195  (39.3066681209, -76.7062702873)  39.306668 -76.706270    2012-12-18  
103196  (39.2958396988, -76.5712467336)  39.295840 -76.571247    2012-12-18  
103197  (39.3436498039, -76.6423081460)  39.343650 -76.642308    2012-12-18  
103198  (39.2958396988, -76.5712467336)  39.295840 -76.571247    2012-12-18  
...                                 ...        ...        ...           ...  
104522  (39.2958396988, -76.5712467336)  39.295840 -76.571247    2012-12-31  
104523  (39.2893323126, -76.6210021717)  39.289332 -76.621002    2012-12-31  
104524  (39.3252613570, -76.5689030474)  39.325261 -76.568903    2012-12-31  
104525  (39.3116837460, -76.6475011849)  39.311684 -76.647501    2012-12-31  
104527  (39.3361054128, -76.6671512244)  39.336105 -76.667151    2

953 entries is a much more manageable size than 63892 entries.

Note: Looking at these entries is useful if we want to analyze crime that occurs in this time frame, but we can't generalize what we see from this data to the whole two-year span. If we wanted to do that, we should instead take a random sample of the data (which should be distributed roughly equally across the time frame). 

In addition, I want to specifically indicate the type of crime with the icon, but there are too many types of incidents to give each their own indicator. In order to make the data comprehensible, we want to group together similar incidents under the same marker. 

First, it makes sense to give the most common incidents their own marker, so let's see which incidents are most commonly reported:

In [6]:
arrest_groups_limited = arrest_table_limited.groupby('incidentOffense')
arrest_groups_limited.count().sort_values('arrest',ascending=False)['arrest']

incidentOffense
87-Narcotics                     308
Unknown Offense                  162
4E-Common Assault                 97
97-Search & Seizure               52
79-Other                          39
24-Towed Vehicle                  35
115-Trespassing                   27
87O-Narcotics (Outside)           26
4B-Agg. Asslt.- Cut               21
4C-Agg. Asslt.- Oth.              19
5A-Burg. Res. (Force)             16
55-Disorderly Person              16
61-Person Wanted On War           12
6C-Larceny- Shoplifting           10
75-Destruct. Of Property           9
55A-Prostitution                   9
54-Armed Person                    9
6D-Larceny- From Auto              8
3AF-Robb Hwy-Firearm               4
5D-Burg. Oth. (Force)              3
111-Protective Order               3
3K-Robb Res. (Ua)                  3
109-Loitering                      3
7A-Stolen Auto                     3
6J-Larceny- Other                  2
20A-Followup                       2
3AK-Robb Hwy-Knife    

Fortunately, it seems that every offense has a crime code if some sort assigned. 

An easy way to categorize all the items would be to give, say, the first 10 entries unique icons, then lumping the other types of crimes into a category 'Other (Categorized)'. To make these distinguishable, we can put each icon type on its own layer. To make it more useful, we can also add an on-hover tooltip that describes the offense in more detail.

(I initially had looked through the [crime codes](https://data.baltimorecity.gov/documents/crime-codes/about) and was going to group them together based on which crimes were most similar, but a majority of the offenses here only take up a small proportion of the total crimes, so it wouldn't be useful to try and group them together in this manner -- i.e. it doesn't help improve clarity.)  

In [9]:
from pandas import isna
map_osm = folium.Map(location=[39.29, -76.61], zoom_start=11)


icons = [
    {'color': 'red', 'icon': 'minus-circle'},
    {'color': 'orange', 'icon': 'question'},
    {'color': 'beige', 'icon': 'hand-rock-o'},
    {'color': 'lightgreen', 'icon': 'search'},
    {'color': 'darkgreen', 'icon': 'ellipsis-h'},
    {'color': 'lightblue', 'icon': 'truck'},
    {'color': 'blue', 'icon': 'blind'},
    {'color': 'darkblue', 'icon': 'minus-circle'},
    {'color': 'purple', 'icon': 'cut'},
    {'color': 'pink', 'icon': 'ban'}
]

icon_other = {'color': 'gray', 'icon': 'ellipsis-v'}

d = dict(tuple(arrest_groups_limited))
offense_list = arrest_groups_limited.count().sort_values('arrest',ascending=False).index

# Note: if there is no given detailed description, I replace it with 'No description'
for offense,icon in zip(offense_list[:len(icons)],icons):
    fg = folium.FeatureGroup(offense)
    d[offense].apply(lambda row: folium.Marker(location=[row['lat'], row['long']], \
                                               tooltip=("No description" if isna(row['chargeDescription']) else row['chargeDescription']), \
                                     icon=folium.Icon(**icon, prefix='fa')).add_to(fg), axis=1)
    fg.add_to(map_osm)
    
# add all the other offenses under 'Other'
fg_other = folium.FeatureGroup('Other (Categorized)')
for offense in offense_list[len(icons):]:
    d[offense].apply(lambda row: folium.Marker(location=[row['lat'], row['long']], \
                                               tooltip=offense + " :: " + \
                                               ("No description" if isna(row['chargeDescription']) else row['chargeDescription']), \
                                     icon=folium.Icon(**icon_other, prefix='fa')).add_to(fg_other), axis=1)
    
fg_other.add_to(map_osm)

folium.LayerControl().add_to(map_osm)


map_osm

### Short Prose Summary
The data shown are all the crimes in the given database that occured in the final two weeks of 2012 (specifically between December 17th and December 31st, inclusive), giving insight into the distribution in that sliver of time. The data is split up into several layers based on the type of offense, as assigned by BPD's crime codes. The ten most common offenses are each given their own layer, including a distinct color pin and icon picture; the rest of them are lumped into a final miscellaneoyus category. (The icons are useful in case the viewer is color blind, as well as helping people get an idea what the offense category means without reading the layer controls.) 

There's a lot of stuff shown on the map, but you can focus on specific crimes (removing the clutter) by changing the layers that are visible in the layer control panel at the top right. In addition, if you hover over each of the specific offenses, you will see extra information about the offense, as well as their provided category if they are labelled as 'Other (Uncategorized)'.  